Add declarative submetric schema for UDF metric

richard-rogers opened the pull request:
## Description

```
udf_schema = DeclarativeSubmetricSchema(
  COLUMN_METRICS + [
    ResolverSpec(column_name="udf1",
                            metrics=COLUMN_METRICS+[MetricSpec(StandardMetric.distribution.value)]),
    ResolverSpec(column_type=Integral,
                            metrics=COLUMN_METRICS+[MetricSpec(StandardMetrics.ints.value)]),
  ]
)

config = UdfMetricConfig(udfs={"udf1": lambda x: 3.14, "udf2": lambda x: 42}, submetric_schema=udf_schema)

dataset_schema = DeclarativeSchema(
  STANDARD_RESOLVERS + [
  ResolverSpec(column_name="udf_input_column",
                           metrics=[MetricSpec(UdfMetric, config))
  ]
)
why.log(df, schema=dataset_schema)
```
This adds a `UdfMetric` to the input column named `udf_input_column`. It will compete the 2 submetrics `udf1` and `udf2` from the value in the `udf_input_column`. The `udf1` output will get the `COLUMN_METRICS` required by WhyLabs plus a distribution metric. Any UDFs that produce integers (in this case, just `udf2`) will get the `COLUMN_METRICS` plus an integer metric.

Also adds a decorator for easy configuration of `UdfMetric`s
```
    @register_metric_udf(col_name="col1")
    def add5(x):
        return x + 5

    @register_metric_udf(col_type=String)
    def upper(x):
        return x.upper()

    dataset_schema = DeclarativeSchema(STANDARD_RESOLVER + generate_udf_schema())
    why.log(df, schema=dataset_schema)
````

## Changes

- Adds `DeclarativeSubmetricSchema <: SubmetricSchema` that uses the `ResolverSpec` and `MetricSpec` to specify submetrics to apply to UDF results
- Adds `@register_metric_udf` decorator for automatically generating submetric schemata

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

richard-rogers committed:
Add declarative submetric schema for UDF metric

FelipeAdachi reviewed (commented):
Can we include a very simple example of how to use this? By example, I mean just adding a few lines to the description of this PR or including a simple test.

richard-rogers committed:
Merge branch 'mainline' into dev/richard/plugin

richard-rogers committed:
Merge branch 'mainline' into dev/richard/plugin

jamie256 commented on python/whylogs/experimental/core/metrics/udf_metric.py:
```diff
@@ -20,37 +12,38 @@
-            metrics["frequent_items"] = StandardMetric.frequent_items.zero(MetricConfig())
 
-        return metrics
+DefaultSchema = DeclarativeSubmetricSchema(STANDARD_RESOLVER)
```
Its surprising and inconsistent styling for `DefaultSchema` to now be an instance of a schema rather than a class.

You could replace DefaultSchema with a factory for creating instances of default schemas?

Or to preserve existing functionality use `functools.partial`:
```suggestion
DefaultSchema = functools.partial(DeclarativeSubmetricSchema, resolvers=STANDARD_RESOLVER)
```

jamie256 commented on python/whylogs/experimental/core/metrics/udf_metric.py:
```diff
@@ -20,37 +12,38 @@
 
     udfs: Dict[str, Callable[[Any], Any]] = field(default_factory=dict)
-    submetric_schema: SubmetricSchema = field(default_factory=DefaultSchema)
+    submetric_schema: SubmetricSchema = field(default_factory=lambda: DefaultSchema)
```
this looks out of place with a lambda here, see suggestions above and we can make this consistent.

jamie256 reviewed (commented):
This could be useful mechanism for plugins, lets clean up the default schema generation, but approach looks fine.

richard-rogers committed:
Merge branch 'mainline' into dev/richard/plugin

richard-rogers committed:
funtools.partial

jamie256 reviewed (dismissed):
LGTM!

richard-rogers committed:
Add decorator to construct schema for UDFs (#1217)

This repository adds a `@register_meric_udf` decorator to auto-generate
dataset & submetric schema for `UdfMetric`.

richard-rogers committed:
Merge branch 'mainline' into dev/richard/plugin

jamie256 commented on python/poetry.lock:
Remove the poetry.lock update from this PR

richard-rogers committed:
mainline lock

jamie256 reviewed (approved):
LGTM!
